impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java

```diff
@@ -396,15 +412,13 @@ private void roleOrderedShutdown(Map<Integer, Server> servers) {
     }
 
     private boolean isController(Integer key) {
-        // TODO this is nasty. We shouldn't need to go via the portAllocator to figure out what a node is
-        // But it is at least testing something meaningful about the configuration
-        return portsAllocator.hasRegisteredPort(Listener.CONTROLLER, key);
+        return key < clusterConfig.numNodes() // dynamically added nodes are always pure brokers
+                && clusterConfig.hasControllerRole(key);
```

Comment:
I think pushing this back to the `clusterConfig` is a good move, but it  moves us back to playing games with nodeId's rather testing for something that is an inherent property of the feature we care about.

Separately I don't see why we need the first half of the condition `key < clusterConfig.numNodes()` it either has the controller role or not. Sprinkling `key < clusterConfig.numNodes()` throughout just defuses the definition of what is and isn't a controller throughout the codebase.

Reply:
cluster config is immutable and encodes the requested number of brokers and controllers. When nodes are added dynamically it doesn't know about them. We might be able to have it reason about `node.id > numNodes()`, but when what does `numNodes()` actually mean? Since it's the `KafkaCluster` that is exposing `add/removeBroker()` it seems reasonable to expect it to be responsible for knowing about the node ids required.

Reply:
I'm not sure if we are agreeing or not. 

I think guessing if something is a controller or not based on its ID is not a good idea. It just bakes in assumptions about the order of operations. The old model of checking the port allocations was an odd choice but at least tested for a property of being a controller or not.

It still strikes me that we should have some way to say given a nodeID what roles does it have. Pushing the question of what does this nodeID do into clusterConfig and thus hiding the nodeID guessing is a decent compromise.

Reply:
I _think_ we agree? With what I've done in this PR then given a `MetadataMode` and the number of brokers and controllers it's completely deterministic what roles any given node id should have. There is no guessing, and the order of operations cannot change the role:  If a node ends up configured with different roles it's simply a bug.